Sources/X509/Verifier/ServerIdentityPolicy.swift

```diff
@@ -0,0 +1,499 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCertificates open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import SwiftASN1
+
+/// A ``VerifierPolicy`` that validates that the leaf certificate is authoritative
+/// for a given hostname or IP address.
+///
+/// This policy is most commonly used to validate the leaf certificate presented by a server
+/// during a TLS handshake.
+///
+/// This policy implements the logic for service validation as specified by
+/// RFC 6125 (https://tools.ietf.org/search/rfc6125), which loosely speaking
+/// defines the common algorithm used for validating that an X.509 certificate
+/// is valid for a given service
+public struct ServerIdentityPolicy {
+    @usableFromInline
+    let serverHostname: String?
+
+    // This field is `var` becuase we lazily convert from String to something more useful, if needed.
+    @usableFromInline
+    var serverIP: LazyIPAddress?
+
+    @usableFromInline
+    var convertedServerHostname: PreparedServerHostname?
+
+    /// Constructs a new ``ServerIdentityPolicy``.
+    ///
+    /// - parameters:
+    ///     - serverHostname: The hostname used to connect to the server.
+    ///     - serverIP: The IP address of the server, if known.
+    @inlinable
+    public init(
+        serverHostname: String?,
+        serverIP: String?
+    ) {
+        self.serverHostname = serverHostname
+        self.serverIP = serverIP.map { .string($0) }
+    }
+}
+
+extension ServerIdentityPolicy: VerifierPolicy {
+    @inlinable
+    public var verifyingCriticalExtensions: [ASN1ObjectIdentifier] {
+        [.X509ExtensionID.subjectAlternativeName]
+    }
+
+    @inlinable
+    public mutating func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain) -> PolicyEvaluationResult {
+        let targetIP = self.serverIP.convert()
+
+        if let serverHostname = self.serverHostname, self.convertedServerHostname == nil {
+            self.convertedServerHostname = PreparedServerHostname(lowercaseASCIIBytes: serverHostname)
+        }
```

Comment:
We are doing another way of caching here as well but forgot to set `serverHostname` to `nil` if the failable  `PreparedServerHostname.init` returns nil. This means we repeatedly are trying to prepare the server hostname if it isn't a valid hostname. I suggest we move this to an init or model the internal state as enum so we are sure we only do initial conversion only once.

Reply:
We can do the enum state, that works for me.